### PR TITLE
Type hints

### LINKED ISSUE #239: Implement type hints
Add type annotations to the code for good practice.

### PR DESCRIPTION
**Pull request by crangelsmith**
Fixes #239 

- Added type hints to most of the codebase
- Add `mypy` pre-commit checks
- Fixed many ambiguities found by `mypy`
- Updated some of the CI workflows (action and python versions)

**Warning**: `mypy` will force type consistency across the codebase, do not be surprised if your pre-commits fail at the beginning while you get used to the structure. All typing is consistent with Python `3.10` and above (a reason my commits are a bit of a mess in this PR :/).

**To do** (another PR): 
- A few functions in the visualisation are still missing type hints. As there is a plan of consolidating these plots I decided to not fully implement, we should ensure we have it in the future. 
- In the same line we should add docstrings where is missing.

### CODE REVIEW

**Review by jolaem — APPROVED** [review]
All tests passing, only one typo noticed.

File: avae/encoders/encoders.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def __init__(
             )
         elif filters is not None and len(filters) != 0:
             if 0 in filters:
-                raise RuntimeError("Filter list cannot contain zeros.")
+                raise RuntimeError("Filter list catorch.nnot contain zeros.")
```
[the remark this answers is not in the document]
Comment (jolaem) [reply avae/encoders/encoders.py:53]: oh, I guess it's a find nn and replace :D

**Review by marjanfamili — APPROVED** [review]
Looks good.

File: avae/encoders/encoders.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (crangelsmith) [reply avae/encoders/encoders.py]: fixed now

**Pull request merged**